File: Views/Components/AsyncImageView.swift
import SwiftUI
import FirebaseStorage
import FirebaseFunctions

struct AsyncImageView: View {
    let imageURL: String?
    let width: CGFloat?
    let height: CGFloat?
    let cornerRadius: CGFloat
    
    @State private var loadedImage: UIImage?
    @State private var isLoading = false
    @State private var hasError = false
    
    init(
        imageURL: String?,
        width: CGFloat? = nil,
        height: CGFloat? = nil,
        cornerRadius: CGFloat = 12
    ) {
        self.imageURL = imageURL
        self.width = width
        self.height = height
        self.cornerRadius = cornerRadius
    }
    
    var body: some View {
        Group {
            if let loadedImage = loadedImage {
                Image(uiImage: loadedImage)
                    .resizable()
                    .aspectRatio(contentMode: .fill)
                    .frame(width: width, height: height)
                    .clipped()
                    .cornerRadius(cornerRadius)
            } else if isLoading {
                Rectangle()
                    .fill(Color.gray.opacity(0.3))
                    .frame(width: width, height: height)
                    .cornerRadius(cornerRadius)
                    .overlay(
                        ProgressView()
                            .scaleEffect(0.8)
                    )
            } else if hasError {
                Rectangle()
                    .fill(Color.gray.opacity(0.2))
                    .frame(width: width, height: height)
                    .cornerRadius(cornerRadius)
                    .overlay(
                        VStack(spacing: 4) {
                            Image(systemName: "photo")
                                .font(.title2)
                                .foregroundColor(.gray)
                            Text("Image non disponible")
                                .font(.caption)
                                .foregroundColor(.gray)
                        }
                    )
            } else {
                Rectangle()
                    .fill(Color.gray.opacity(0.2))
                    .frame(width: width, height: height)
                    .cornerRadius(cornerRadius)
                    .overlay(
                        Image(systemName: "photo")
                            .font(.title2)
                            .foregroundColor(.gray)
                    )
            }
        }
        .onAppear {
            loadImageIfNeeded()
        }
        .onChange(of: imageURL) { oldValue, newValue in
            loadImageIfNeeded()
        }
    }
    
    private func loadImageIfNeeded() {
        guard let imageURL = imageURL,
              !imageURL.isEmpty,
              loadedImage == nil,
              !isLoading else {
            print("🖼️ AsyncImageView: loadImageIfNeeded - Conditions non remplies")
            print("🖼️ AsyncImageView: - imageURL: \(imageURL ?? "nil")")
            print("🖼️ AsyncImageView: - isEmpty: \(imageURL?.isEmpty ?? true)")
            print("🖼️ AsyncImageView: - loadedImage: \(loadedImage != nil ? "présente" : "nil")")
            print("🖼️ AsyncImageView: - isLoading: \(isLoading)")
            return
        }
        
        print("🖼️ AsyncImageView: Tentative de chargement de l'image: \(imageURL)")
        loadImage(from: imageURL)
    }
    
    private func loadImage(from urlString: String) {
        print("🖼️ AsyncImageView: loadImage démarré pour: \(urlString)")
        isLoading = true
        hasError = false
        
        Task {
            do {
                let image = try await loadImageFromFirebase(urlString: urlString)
                
                print("✅ AsyncImageView: Image chargée avec succès pour: \(urlString)")
                await MainActor.run {
                    self.loadedImage = image
                    self.isLoading = false
                    self.hasError = false
                }
            } catch {
                print("❌ AsyncImageView: Erreur chargement image pour \(urlString): \(error)")
                
                await MainActor.run {
                    self.isLoading = false
                    self.hasError = true
                }
            }
        }
    }
    
    private func loadImageFromFirebase(urlString: String) async throws -> UIImage {
        print("🖼️ AsyncImageView: loadImageFromFirebase appelé avec: \(urlString)")
        
        // 1. NOUVEAU: Vérifier le cache d'abord
        if let cachedImage = ImageCacheService.shared.getCachedImage(for: urlString) {
            print("🖼️ AsyncImageView: Image trouvée en cache ! Aucun téléchargement nécessaire")
            return cachedImage
        }
        
        // 2. Si pas en cache, télécharger selon le type d'URL
        let downloadedImage: UIImage
        
        if urlString.contains("firebasestorage.googleapis.com") {
            // 🔧 NOUVEAU: Utiliser Cloud Function pour TOUTES les images Firebase Storage
            print("🖼️ AsyncImageView: URL Firebase Storage détectée - Utilisation Cloud Function...")
            downloadedImage = try await loadFromFirebaseStorageViaCloudFunction(urlString: urlString)
        } else if urlString.hasPrefix("placeholder_image_") {
            // Placeholder
            print("🖼️ AsyncImageView: Image placeholder détectée, génération d'une image par défaut")
            downloadedImage = try await createPlaceholderImage()
        } else {
            // URL normale
            print("🖼️ AsyncImageView: Tentative de chargement URL normale: \(urlString)")
            guard let url = URL(string: urlString) else {
                print("❌ AsyncImageView: URL invalide: \(urlString)")
                throw AsyncImageError.invalidData
            }
            
            let (data, _) = try await URLSession.shared.data(from: url)
            
            guard let image = UIImage(data: data) else {
                print("❌ AsyncImageView: Impossible de créer UIImage depuis les données")
                throw AsyncImageError.invalidData
            }
            
            print("✅ AsyncImageView: Image chargée depuis URL: \(urlString)")
            downloadedImage = image
        }
        
        // 3. NOUVEAU: Mettre l'image téléchargée en cache
        ImageCacheService.shared.cacheImage(downloadedImage, for: urlString)
        print("🖼️ AsyncImageView: Image mise en cache pour utilisation future")
        
        return downloadedImage
    }
    
    // 🔧 NOUVELLE MÉTHODE: Utiliser Cloud Function pour contourner les règles Firebase Storage
    private func loadFromFirebaseStorageViaCloudFunction(urlString: String) async throws -> UIImage {
        print("🔧 AsyncImageView: Tentative de chargement via Cloud Function pour: \(urlString)")
        
        // Extraire le chemin du fichier depuis l'URL
        guard let urlMatch = urlString.range(of: "/o/(.+?)\\?", options: .regularExpression),
              let encodedPath = String(urlString[urlMatch]).components(separatedBy: "/o/").last?.components(separatedBy: "?").first else {
            print("❌ AsyncImageView: Impossible d'extraire le chemin du fichier depuis l'URL")
            // Fallback vers la méthode directe
            return try await loadFromFirebaseStorageDirect(urlString: urlString)
        }
        
        let filePath = encodedPath.removingPercentEncoding ?? encodedPath
        print("🔧 AsyncImageView: Chemin extrait: \(filePath)")
        
        // Déterminer le type d'image et l'ID utilisateur
        if filePath.hasPrefix("profile_images/") {
            // Image de profil - extraire l'ID utilisateur
            let pathComponents = filePath.components(separatedBy: "/")
            if pathComponents.count >= 2 {
                let userId = pathComponents[1]
                print("🔧 AsyncImageView: Image de profil détectée pour utilisateur: \(userId)")
                return try await loadProfileImageViaCloudFunction(userId: userId)
            }
        } else if filePath.hasPrefix("journal_images/") {
            // Image du journal - utiliser URL signée générique
            print("🔧 AsyncImageView: Image du journal détectée")
            return try await loadImageViaSignedURL(filePath: filePath)
        }
        
        // Fallback vers la méthode directe si on ne peut pas identifier le type
        print("🔧 AsyncImageView: Type d'image non identifié, fallback vers méthode directe")
        return try await loadFromFirebaseStorageDirect(urlString: urlString)
    }
    
    // Charger image de profil via Cloud Function existante
    private func loadProfileImageViaCloudFunction(userId: String) async throws -> UIImage {
        return try await withCheckedThrowingContinuation { continuation in
            let functions = Functions.functions()
            
            functions.httpsCallable("getPartnerProfileImage").call(["partnerId": userId]) { result, error in
                if let error = error {
                    print("❌ AsyncImageView: Erreur Cloud Function getPartnerProfileImage: \(error)")
                    continuation.resume(throwing: error)
                    return
                }
                
                guard let data = result?.data as? [String: Any],
                      let success = data["success"] as? Bool,
                      success,
                      let signedUrl = data["imageUrl"] as? String else {
                    print("❌ AsyncImageView: Réponse Cloud Function invalide")
                    continuation.resume(throwing: AsyncImageError.loadingFailed)
                    return
                }
                
                print("✅ AsyncImageView: URL signée obtenue via Cloud Function: \(signedUrl)")
                
                // Charger l'image depuis l'URL signée
                Task {
                    do {
                        guard let url = URL(string: signedUrl) else {
                            continuation.resume(throwing: AsyncImageError.invalidData)
                            return
                        }
                        
                        let (imageData, _) = try await URLSession.shared.data(from: url)
                        
                        guard let image = UIImage(data: imageData) else {
                            continuation.resume(throwing: AsyncImageError.invalidData)
                            return
                        }
                        
                        print("✅ AsyncImageView: Image de profil chargée via Cloud Function")
                        continuation.resume(returning: image)
                    } catch {
                        print("❌ AsyncImageView: Erreur chargement depuis URL signée: \(error)")
                        continuation.resume(throwing: error)
                    }
                }
            }
        }
    }
    
    // Nouvelle méthode pour charger via URL signée générique
    private func loadImageViaSignedURL(filePath: String) async throws -> UIImage {
        return try await withCheckedThrowingContinuation { continuation in
            let functions = Functions.functions()
            
            functions.httpsCallable("getSignedImageURL").call(["filePath": filePath]) { result, error in
                if let error = error {
                    print("❌ AsyncImageView: Erreur Cloud Function getSignedImageURL: \(error)")
                    // Fallback vers méthode directe
                    Task {
                        do {
                            let image = try await self.loadFromFirebaseStorageDirect(urlString: "gs://love2love-26164.firebasestorage.app/\(filePath)")
                            continuation.resume(returning: image)
                        } catch {
                            continuation.resume(throwing: error)
                        }
                    }
                    return
                }
                
                guard let data = result?.data as? [String: Any],
                      let success = data["success"] as? Bool,
                      success,
                      let signedUrl = data["signedUrl"] as? String else {
                    print("❌ AsyncImageView: Réponse Cloud Function getSignedImageURL invalide")
                    // Fallback vers méthode directe
                    Task {
                        do {
                            let image = try await self.loadFromFirebaseStorageDirect(urlString: "gs://love2love-26164.firebasestorage.app/\(filePath)")
                            continuation.resume(returning: image)
                        } catch {
                            continuation.resume(throwing: error)
                        }
                    }
                    return
                }
                
                print("✅ AsyncImageView: URL signée obtenue pour image du journal: \(signedUrl)")
                
                // Charger l'image depuis l'URL signée
                Task {
                    do {
                        guard let url = URL(string: signedUrl) else {
                            continuation.resume(throwing: AsyncImageError.invalidData)
                            return
                        }
                        
                        let (imageData, _) = try await URLSession.shared.data(from: url)
                        
                        guard let image = UIImage(data: imageData) else {
                            continuation.resume(throwing: AsyncImageError.invalidData)
                            return
                        }
                        
                        print("✅ AsyncImageView: Image du journal chargée via URL signée")
                        continuation.resume(returning: image)
                    } catch {
                        print("❌ AsyncImageView: Erreur chargement depuis URL signée: \(error)")
                        continuation.resume(throwing: error)
                    }
                }
            }
        }
    }
    
    // Méthode directe en fallback (ancienne méthode)
    private func loadFromFirebaseStorageDirect(urlString: String) async throws -> UIImage {
        print("🔧 AsyncImageView: Fallback vers méthode directe Firebase Storage")
        return try await withCheckedThrowingContinuation { continuation in
            let storageRef = Storage.storage().reference(forURL: urlString)
            
            // Télécharger l'image avec une taille maximale de 10MB
            storageRef.getData(maxSize: 10 * 1024 * 1024) { data, error in
                if let error = error {
                    print("❌ AsyncImageView: Erreur téléchargement Firebase Storage direct: \(error)")
                    continuation.resume(throwing: error)
                    return
                }
                
                guard let data = data,
                      let image = UIImage(data: data) else {
                    print("❌ AsyncImageView: Données Firebase Storage invalides")
                    continuation.resume(throwing: AsyncImageError.invalidData)
                    return
                }
                
                print("✅ AsyncImageView: Image chargée depuis Firebase Storage direct")
                continuation.resume(returning: image)
            }
        }
    }
    
    private func createPlaceholderImage() async throws -> UIImage {
        return await withCheckedContinuation { continuation in
            let size = CGSize(width: 300, height: 200)
            let renderer = UIGraphicsImageRenderer(size: size)
            
            let image = renderer.image { context in
                // Fond dégradé
                let colors = [UIColor.systemPink.withAlphaComponent(0.3).cgColor,
                             UIColor.systemPurple.withAlphaComponent(0.3).cgColor]
                let gradient = CGGradient(colorsSpace: CGColorSpaceCreateDeviceRGB(),
                                        colors: colors as CFArray,
                                        locations: nil)!
                
                context.cgContext.drawLinearGradient(gradient,
                                                   start: CGPoint(x: 0, y: 0),
                                                   end: CGPoint(x: size.width, y: size.height),
                                                   options: [])
                
                // Icône cœur au centre
                let heartSize: CGFloat = 40
                let heartRect = CGRect(x: (size.width - heartSize) / 2,
                                     y: (size.height - heartSize) / 2,
                                     width: heartSize,
                                     height: heartSize)
                
                UIColor.systemPink.setFill()
                let heartPath = UIBezierPath()
                let centerX = heartRect.midX
                let centerY = heartRect.midY
                
                // Dessiner un cœur simple
                heartPath.move(to: CGPoint(x: centerX, y: centerY + 10))
                heartPath.addCurve(to: CGPoint(x: centerX - 15, y: centerY - 5),
                                 controlPoint1: CGPoint(x: centerX - 5, y: centerY + 5),
                                 controlPoint2: CGPoint(x: centerX - 15, y: centerY))
                heartPath.addCurve(to: CGPoint(x: centerX, y: centerY - 15),
                                 controlPoint1: CGPoint(x: centerX - 15, y: centerY - 10),
                                 controlPoint2: CGPoint(x: centerX - 5, y: centerY - 15))
                heartPath.addCurve(to: CGPoint(x: centerX + 15, y: centerY - 5),
                                 controlPoint1: CGPoint(x: centerX + 5, y: centerY - 15),
                                 controlPoint2: CGPoint(x: centerX + 15, y: centerY - 10))
                heartPath.addCurve(to: CGPoint(x: centerX, y: centerY + 10),
                                 controlPoint1: CGPoint(x: centerX + 15, y: centerY),
                                 controlPoint2: CGPoint(x: centerX + 5, y: centerY + 5))
                heartPath.close()
                heartPath.fill()
            }
            
            continuation.resume(returning: image)
        }
    }
}

enum AsyncImageError: LocalizedError {
    case invalidData
    case loadingFailed
    
    var errorDescription: String? {
        switch self {
        case .invalidData:
            return "Données d'image invalides"
        case .loadingFailed:
            return "Échec du chargement de l'image"
        }
    }
}

#Preview {
    VStack(spacing: 20) {
        AsyncImageView(imageURL: nil, width: 200, height: 150)
        AsyncImageView(imageURL: "", width: 200, height: 150)
        AsyncImageView(imageURL: "invalid-url", width: 200, height: 150)
    }
    .padding()
} 

File: Views/Components/AsyncImageView.kt
// AsyncImageView.kt — Jetpack Compose + Firebase (Storage, Functions, Auth)
// ---------------------------------------------------------------
// ✅ Points clés par rapport à ta version Swift
// - Traductions via strings.xml (ex: stringResource(R.string.image_not_available))
// - Cache mémoire local (LruCache) comme ImageCacheService.shared
// - Chargement via Cloud Functions pour:
//     • getPartnerProfileImage(partnerId)
//     • getSignedImageURL(filePath)
// - Fallback direct Firebase Storage (getReferenceFromUrl)
// - Placeholders / état de chargement / erreur, coins arrondis
// - Logs sécurisés (pas d’URL complète dans les logs)
// ---------------------------------------------------------------
// Dépendances Gradle (module app) à vérifier :
//
// implementation(platform("com.google.firebase:firebase-bom:33.4.0"))
// implementation("com.google.firebase:firebase-storage-ktx")
// implementation("com.google.firebase:firebase-functions-ktx")
// implementation("com.google.firebase:firebase-auth-ktx")
// implementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:1.8.1")
// implementation("org.jetbrains.kotlinx:kotlinx-coroutines-play-services:1.8.1")
// implementation("androidx.compose.ui:ui:1.7.2")
// implementation("androidx.compose.material3:material3:1.3.0")
// implementation("androidx.compose.material:material-icons-extended:1.7.2")
// ---------------------------------------------------------------
// strings.xml (extrait — à placer dans res/values/strings.xml)
// <resources>
//     <string name="image_not_available">Image non disponible</string>
// </resources>
// ---------------------------------------------------------------

package com.love2loveapp.core.ui

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Image
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.ImageBitmap
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.love2loveapp.core.R
import com.love2loveapp.core.utils.ImageCacheService
import com.love2loveapp.core.utils.FirebaseImageRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

@Composable
fun AsyncImageView(
    imageUrl: String?,
    width: Dp? = null,
    height: Dp? = null,
    cornerRadius: Dp = 12.dp,
) {
    val TAG = "AsyncImageView"
    val context = LocalContext.current

    var loadedImage by remember(imageUrl) { mutableStateOf<ImageBitmap?>(null) }
    var isLoading by remember(imageUrl) { mutableStateOf(false) }
    var hasError by remember(imageUrl) { mutableStateOf(false) }

    // Détermine le modifier en fonction des dimensions fournies
    var modifier = Modifier.clip(RoundedCornerShape(cornerRadius))
    modifier = when {
        width != null && height != null -> modifier.size(width, height)
        width != null -> modifier.width(width)
        height != null -> modifier.height(height)
        else -> modifier
    }

    // UI
    Box(modifier = modifier) {
        when {
            loadedImage != null -> {
                Image(
                    bitmap = loadedImage!!,
                    contentDescription = null,
                    contentScale = ContentScale.Crop,
                    modifier = Modifier.fillMaxSize()
                )
            }
            isLoading -> {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(MaterialTheme.colorScheme.surfaceVariant),
                    contentAlignment = Alignment.Center
                ) {
                    CircularProgressIndicator()
                }
            }
            hasError -> {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(MaterialTheme.colorScheme.surfaceVariant),
                    contentAlignment = Alignment.Center
                ) {
                    androidx.compose.foundation.layout.Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Icon(
                            imageVector = Icons.Outlined.Image,
                            contentDescription = null
                        )
                        Text(
                            text = stringResource(id = R.string.image_not_available),
                            style = MaterialTheme.typography.labelSmall
                        )
                    }
                }
            }
            else -> {
                // Placeholder initial (sans erreur et sans chargement)
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(MaterialTheme.colorScheme.surfaceVariant),
                    contentAlignment = Alignment.Center
                ) {
                    Icon(
                        imageVector = Icons.Outlined.Image,
                        contentDescription = null
                    )
                }
            }
        }
    }

    // Chargement côté effet
    val scope = rememberCoroutineScope()
    LaunchedEffect(imageUrl) {
        if (imageUrl.isNullOrBlank()) {
            Log.d(TAG, "loadImageIfNeeded: URL manquante — rien à charger")
            loadedImage = null
            isLoading = false
            hasError = false
            return@LaunchedEffect
        }

        if (loadedImage != null || isLoading) {
            Log.d(TAG, "loadImageIfNeeded: déjà chargé ou en cours")
            return@LaunchedEffect
        }

        isLoading = true
        hasError = false
        Log.d(TAG, "Tentative de chargement de l'image: [URL MASQUÉE]")

        scope.launch {
            try {
                // 1) Cache
                ImageCacheService.getCached(imageUrl)?.let { bmp ->
                    Log.d(TAG, "Image trouvée en cache — pas de téléchargement")
                    loadedImage = bmp.asImageBitmap()
                    isLoading = false
                    hasError = false
                    return@launch
                }

                // 2) Placeholder géré côté UI si préfixe
                if (imageUrl.startsWith("placeholder_image_")) {
                    Log.d(TAG, "Placeholder détecté — aucun téléchargement nécessaire")
                    loadedImage = null // on laisse l'UI afficher un placeholder neutre
                    isLoading = false
                    hasError = false
                    return@launch
                }

                // 3) Téléchargement selon type d'URL
                val bmp = withContext(Dispatchers.IO) {
                    FirebaseImageRepository.loadBitmapSmart(urlString = imageUrl)
                }

                // 4) Cache + rendu
                ImageCacheService.put(imageUrl, bmp)
                loadedImage = bmp.asImageBitmap()
                isLoading = false
                hasError = false
                Log.d(TAG, "Image chargée avec succès")
            } catch (t: Throwable) {
                Log.e(TAG, "Erreur de chargement [URL MASQUÉE]", t)
                isLoading = false
                hasError = true
            }
        }
    }
}

// ---------------------------------------------------------------
// FirebaseImageRepository — logique de chargement (Cloud Functions / Storage / HTTP)
// ---------------------------------------------------------------
package com.love2loveapp.core.utils

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log
import com.google.android.gms.tasks.Tasks
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.functions.FirebaseFunctions
import com.google.firebase.storage.FirebaseStorage
import kotlinx.coroutines.tasks.await
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.net.HttpURLConnection
import java.net.URL

object FirebaseImageRepository {
    private const val TAG = "FirebaseImageRepo"

    // ⚠️ Adapte ce bucket si besoin pour les fallbacks "gs://" quand on n'a que le chemin
    private const val FALLBACK_GS_BUCKET = "gs://love2love-26164.firebasestorage.app"

    suspend fun loadBitmapSmart(urlString: String): Bitmap {
        Log.d(TAG, "loadBitmapSmart: début (URL masquée)")

        return when {
            urlString.contains("firebasestorage.googleapis.com") -> {
                // Utiliser Cloud Functions selon le type d'image
                loadFromFirebaseStorageViaCloudFunction(urlString)
            }
            urlString.startsWith("gs://") -> {
                loadFromFirebaseStorageDirect(urlString)
            }
            urlString.startsWith("http://") || urlString.startsWith("https://") -> {
                loadFromHttp(urlString)
            }
            else -> {
                Log.w(TAG, "URL non reconnue — tentative Storage direct")
                loadFromFirebaseStorageDirect(urlString)
            }
        }
    }

    // --------------------- Sous-fonctions ---------------------

    private suspend fun loadFromFirebaseStorageViaCloudFunction(urlString: String): Bitmap {
        Log.d(TAG, "loadFromFirebaseStorageViaCloudFunction: URL Firebase détectée")
        val filePath = extractFilePathFromDownloadUrl(urlString)
        if (filePath == null) {
            Log.w(TAG, "Impossible d'extraire le chemin — fallback direct")
            return loadFromFirebaseStorageDirect(urlString)
        }

        return when {
            filePath.startsWith("profile_images/") -> {
                val parts = filePath.split('/')
                val imageUserId = parts.getOrNull(1)
                val currentUserId = FirebaseAuth.getInstance().currentUser?.uid

                if (imageUserId != null && currentUserId != null && imageUserId == currentUserId) {
                    Log.d(TAG, "Image du user courant → download direct")
                    loadFromFirebaseStorageDirect(urlString)
                } else {
                    Log.d(TAG, "Image d'un partenaire → Cloud Function getPartnerProfileImage")
                    loadProfileImageViaCloudFunction(imageUserId)
                }
            }
            filePath.startsWith("journal_images/") -> {
                Log.d(TAG, "Image journal → Cloud Function getSignedImageURL")
                loadImageViaSignedURL(filePath)
            }
            else -> {
                Log.d(TAG, "Type non identifié → fallback direct")
                loadFromFirebaseStorageDirect(urlString)
            }
        }
    }

    private suspend fun loadProfileImageViaCloudFunction(userId: String?): Bitmap {
        require(!userId.isNullOrBlank()) { "userId requis pour getPartnerProfileImage" }
        val data = mapOf("partnerId" to userId)
        val result = FirebaseFunctions.getInstance()
            .getHttpsCallable("getPartnerProfileImage")
            .call(data)
            .await()

        val map = result.data as? Map<*, *> ?: error("Réponse invalide")
        val success = map["success"] as? Boolean ?: false
        val signedUrl = map["imageUrl"] as? String
        require(success && !signedUrl.isNullOrBlank()) { "Cloud Function getPartnerProfileImage a échoué" }

        return loadFromHttp(signedUrl!!)
    }

    private suspend fun loadImageViaSignedURL(filePath: String): Bitmap {
        val data = mapOf("filePath" to filePath)
        return try {
            val result = FirebaseFunctions.getInstance()
                .getHttpsCallable("getSignedImageURL")
                .call(data)
                .await()
            val map = result.data as? Map<*, *> ?: error("Réponse invalide")
            val success = map["success"] as? Boolean ?: false
            val signedUrl = map["signedUrl"] as? String
            if (success && !signedUrl.isNullOrBlank()) {
                loadFromHttp(signedUrl!!)
            } else {
                Log.w(TAG, "Réponse Cloud Function invalide → fallback direct")
                loadFromFirebaseStorageDirect("$FALLBACK_GS_BUCKET/$filePath")
            }
        } catch (t: Throwable) {
            Log.e(TAG, "getSignedImageURL a échoué → fallback direct", t)
            loadFromFirebaseStorageDirect("$FALLBACK_GS_BUCKET/$filePath")
        }
    }

    private suspend fun loadFromFirebaseStorageDirect(urlOrGs: String): Bitmap {
        Log.d(TAG, "Download direct depuis Firebase Storage (URL masquée)")
        val storageRef = FirebaseStorage.getInstance().getReferenceFromUrl(urlOrGs)
        val bytes = storageRef.getBytes(10L * 1024L * 1024L).await()
        return BitmapFactory.decodeByteArray(bytes, 0, bytes.size)
    }

    private fun extractFilePathFromDownloadUrl(downloadUrl: String): String? {
        // Cherche entre "/o/" et "?"
        val regex = "/o/([^?]+)".toRegex()
        val match = regex.find(downloadUrl) ?: return null
        val encoded = match.groupValues.getOrNull(1) ?: return null
        return try {
            java.net.URLDecoder.decode(encoded, "UTF-8")
        } catch (_: Throwable) {
            encoded
        }
    }

    private suspend fun loadFromHttp(urlString: String): Bitmap {
        Log.d(TAG, "HTTP download (URL masquée)")
        val url = URL(urlString)
        val connection = (url.openConnection() as HttpURLConnection).apply {
            connectTimeout = 10000
            readTimeout = 15000
            instanceFollowRedirects = true
        }
        return connection.inputStream.use { input ->
            BitmapFactory.decodeStream(input)
        }
    }
}

// ---------------------------------------------------------------
// ImageCacheService — cache mémoire (équivalent du service Swift)
// ---------------------------------------------------------------
package com.love2loveapp.core.utils

import android.graphics.Bitmap
import android.util.LruCache

object ImageCacheService {
    private val cache: LruCache<String, Bitmap>

    init {
        val maxKb = (Runtime.getRuntime().maxMemory() / 1024).toInt()
        val cacheSize = maxKb / 8 // ~12.5% de la RAM
        cache = object : LruCache<String, Bitmap>(cacheSize) {
            override fun sizeOf(key: String, value: Bitmap): Int = value.byteCount / 1024
        }
    }

    fun getCached(key: String): Bitmap? = cache.get(key)
    fun put(key: String, bitmap: Bitmap) { cache.put(key, bitmap) }
    fun clear() { cache.evictAll() }
}

// ---------------------------------------------------------------
// Exemple d'utilisation (Composable)
// ---------------------------------------------------------------
package com.love2loveapp.feature.preview

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.love2loveapp.core.ui.AsyncImageView

@Composable
fun AsyncImageViewPreview() {
    Column(Modifier.padding(16.dp)) {
        AsyncImageView(imageUrl = null, width = 200.dp, height = 150.dp)
        AsyncImageView(imageUrl = "", width = 200.dp, height = 150.dp)
        AsyncImageView(imageUrl = "invalid-url", width = 200.dp, height = 150.dp)
    }
}
